### ISSUE TITLE
Tracing and Scripting nn.Conv2d shows different op in the graph

### ISSUE THREAD

**Issue by wanchaol** [pytorch/pytorch#18975]
> I have a simple model with nn.Conv2d. When i trace, graphs shows below aten op 
>%input : Double(1, 20, 24, 24) = aten::_convolution(%input.1, %1, %2, %5, %8, %11, %12, %15, >%16, %17, %18, %19), scope: testModule/Conv2d[conv1]

>When using script front end, it shows :
> %274 : Tensor = aten::conv2d(%ret, %1, %2, %272, %271, %273, %21)

This happens because in variableType we only record the trace for `aten::_convolution` but not any other conv ops like `aten::conv2d` etc. 

I think it's better for us to make scripting and tracing frontend more consistent in ops that emit so that we could do optimizations together to both.

**Comment by zdevito** [comment]
This is due to this hack:

```
# These functions we don't want to record for tracing, because we always want
# to trace their constituent parts.  This is a temporary hack in lieue
# of proper scopes, where subsequent compilation passes can ask for the unfolding
# on demand.  Only concrete ATen methods can be disabled this way; it will have
# NO EFFECT otherwise.
DONT_RECORD_TRACE = {
    'convolution', 'conv1d', 'conv2d', 'conv3d', 'conv_transpose1d',
    'conv_transpose2d', 'conv_transpose3d', 'lstm_cell', 'gru_cell',
    'rnn_tanh_cell', 'rnn_relu_cell', 'linear',
    # FIXME: figure out a better way when we support sparse tensors in jit
    '_coalesced_',
}
```

`_convolution` is the more fundamental op, and we should be recording that one. However, all the definitions are in C++ forwarding conv2d to _convolution.'

The right fix is to remove DONT_RECORD_TRACE, and instead to have definitions of these functions in builtin_functions.cpp that are inserted in place of these functions early in the compilation process. This will work for both scripting and tracing in the same way.

**Comment by wanchaol** [comment]
@zdevito Thanks for suggestion! I started working on a fix as you suggested. One thing I noticed in this approach is that, builtin_functions.cpp compiling does not really know the runtime backend that use could choose. Taking a example of the [conv2d call _convolution](https://github.com/pytorch/pytorch/blob/1e78252de70c1e6361df94ab9b91de06320f253e/aten/src/ATen/native/Convolution.cpp#L342) op, in C++ it can initialize an context and know if `userEnablecuDNN`, but it would be hard to get those informations in TorchScript. This backend selection or environment discovery also happens in autodiff symbolic_script, which I think it's difficult to get in non-C++ environment. Shall we have `Context` as first class type in TorchScript so that we could fetch those informations?

**Comment by xiaosun86** [comment]
@zdevito Can you also comment on the differentiability of conv2d? I checked the forward graph of a simple MNIST model:


class Net(torch.jit.ScriptModule): 

    def __init__(self):
        super(Net, self).__init__() 
        self.conv1 = torch.jit.trace(nn.Conv2d(1, 10, kernel_size=5), torch.randn(1,1,28,28) )
        self.conv2 = torch.jit.trace(nn.Conv2d(10, 20, kernel_size=5), torch.randn(1, 10, 28,28))
        self.fc1 = torch.jit.trace(nn.Linear(320, 50), torch.randn(49,320))
        self.fc2 = torch.jit.trace(nn.Linear(50, 10), torch.rand(49,50))

    @torch.jit.script_method
    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = F.relu(F.max_pool2d(self.conv2(x), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x, dim=1)

which calls aten::_convolution but not puts it into a  `prim::DifferentiableGraph`, and thus no backward_graph is generated for the two convolutions in the model.
Thanks!

**Comment by zdevito** [comment]
I do not think we have a formula for `_convolution` but that doesn't mean that we can't differentiate it, just that we fallback to the standard autograd logic.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
